### PR TITLE
Optimization of I/O path of python interface

### PR DESCRIPTION
**Pull request by cs-ally**

### CODE REVIEW

**Comment by Baunsgaard** [comment]
Hi @cs-ally 

Thanks for the PR.

Before I go into details on the PR, I suggest you address the failing test cases in the Python API. The log indicates that they relate to your implementation.

Best regards
Sebastian

**Comment by cs-ally** [comment]
processing_times_comparison
Processing time improvement visualisation

**Comment by Baunsgaard** [comment]
> processing_times_comparison Processing time improvement visualisation

great, good results. The plot can be improved by marking relative speedup as in 5x faster etc.
It would be good if we can have the old implementation as a fall back solution in the cases with many columns, to not have any speed downs.

**Comment by cs-ally** [comment]
Screenshot 2024-07-27 at 22 19 52

File: src/main/java/org/apache/sysds/runtime/util/Py4jConverterUtils.java [hunk 1]
```diff
@@ -114,6 +120,82 @@ public static MatrixBlock convertPy4JArrayToMB(byte[] data, int rlen, int clen,
 		return mb;
 	}
 
+	public static Array<?> convert(byte[] data, int numElements, Types.ValueType valueType) {
```
Comment (Baunsgaard) [inline src/main/java/org/apache/sysds/runtime/util/Py4jConverterUtils.java:123]: we need to add some internal tests (java only) for this function.
however the implementation does look fine. There are some minor improvements possible if we do not cast the array types inside the inner loops but assign the cast arrays outside the loop and then use that allocation.

File: src/main/python/systemds/utils/converters.py [hunk 2]
```diff
@@ -116,20 +121,46 @@ def pandas_to_frame_block(sds, pd_df: pd.DataFrame):
...
-        for j, col_name in enumerate(col_names):
-            col_data = pd_df[col_name].fillna("").to_numpy(dtype=str)
-            for i in range(col_data.shape[0]):
-                if col_data[i]:
-                    j_dataArray[i][j] = col_data[i]
-        fb = jc_FrameBlock(j_valueTypeArray, j_colNameArray, j_dataArray)
-
-        return fb
+        
+        if rows > 4:
```
Comment (Baunsgaard) [inline src/main/python/systemds/utils/converters.py:125]: add a comment for why we make this differentiation.

File: src/main/python/tests/iotests/test_io_pandas_systemds.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        print("result_df:")
+        result_df['C2'] = result_df['C2'].astype(int)
+        print(result_df)
+        print("self.df:")
+        print(self.df)   
+        self.assertTrue(isinstance(result_df, pd.DataFrame))
+        self.assertTrue(self.df.equals(result_df))
+
+if __name__ == "__main__":
+    unittest.main(exit=False)
```
Comment (Baunsgaard) [inline src/main/python/tests/iotests/test_io_pandas_systemds.py:89]: newline in the end.

**Review by Baunsgaard — COMMENTED** [review]
Overall LGTM, we just need to add an internal Java test, to make the code coverage happy.

File: src/main/java/org/apache/sysds/runtime/util/Py4jConverterUtils.java (hunk 1 above)
Comment (cs-ally) [reply src/main/java/org/apache/sysds/runtime/util/Py4jConverterUtils.java:123]: I wasn’t able to find existing tests for the Py4jConverterUtils class. Could you please advise where I should add new tests?

Comment (Baunsgaard) [reply src/main/java/org/apache/sysds/runtime/util/Py4jConverterUtils.java:123]: I don't think we have any.
You will have to make some from scratch.
It should be fairly simple to allocate byte[] and then call the methods. 

Please put the new tests in 

src/test/java/org/apache/sysds/test/component/frame/array

**Comment by Baunsgaard** [comment]
Merging in #2065  , 
Thanks for the contribution @cs-ally 
While merging I fix the remaining issues, with tabs vs spaces and such.

**Pull request closed without merging**